plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -596,27 +601,152 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String bucketName = getBucketName(uri);
+        String key = keyFromPath(path);
+        ObjectMetadata s3ObjectMetadata = getS3ObjectMetadata(path, bucketName, key);
+        ObjectMetadata s3DirectoryMetadata = null;
+        if (!key.isEmpty()) {
+            s3DirectoryMetadata = getS3ObjectMetadata(path, bucketName, key + PATH_SEPARATOR);
+        }
+
+        if (recursive) {
+            boolean deleted;
+            if (key.endsWith(DIRECTORY_SUFFIX)) {
+                Path strippedPath = new Path(stripDirectorySuffix(path.toString()));
+                deleted = deletePrefix(strippedPath);
+                deleted = deleteObject(stripDirectorySuffix(key) + PATH_SEPARATOR) || deleted;
+            }
+            else {
+                deleted = deletePrefix(path);
+            }
+            if (s3ObjectMetadata != null) {
+                deleted = deleteObject(key) || deleted;
             }
+            if (s3DirectoryMetadata != null) {
+                String directoryKey = key + PATH_SEPARATOR;
+                deleted = deleteObject(directoryKey) || deleted;
+            }
+            return deleted;
         }
-        catch (FileNotFoundException e) {
+        else {
+            if (key.endsWith(DIRECTORY_SUFFIX)) {
+                Path strippedPath = new Path(stripDirectorySuffix(path.toString()));
+                if (listPath(strippedPath, OptionalInt.of(1), ListingMode.SHALLOW_ALL).hasNext()) {
+                    throw new IOException("Directory " + path + " is not empty");
+                }
+                return deleteObject(stripDirectorySuffix(key) + PATH_SEPARATOR);
```

Comment:
If we end up here and `s3ObjectMetadata != null`, then shouldn't we be issuing a `deleteObject(key)` instead? We know that `s3://bucket/directory_$folder$` exists and is the "marker" that indicates the path `s3://bucket/directory` should be interpreted as a folder- but the S3 object key that needs deletion does actually have `_$folder$` suffixed.